Fix Issue #809 - Add Error Logging for Development Chaincode

juliancarrivick-ibm opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
## Description

Add Error message when a user attempts an action other than deploy when they are running their own chaincode and in Developer Mode
## Motivation and Context

<!-- Why is this change required? What problem does it solve? -->

<!-- If it fixes an open issue, please link to the issue here. -->

Fixes #809 
The existing log messages were misleading for developers when they forgot to deploy chaincode before invoking or querying. This will flag that situation in the logs.
## How Has This Been Tested?

<!-- If this PR does not contain a new test case, explain why. -->

<!-- Describe in detail how you tested your changes. -->

I have manually verified that an error message is generated when invoke is called on Chaincode that has not been deployed yet:

```
07:05:49.481 [chaincode] Launch -> DEBU 053 Container not in READY state(established)...send init/ready
07:05:49.483 [chaincode] Launch -> ERRO 054 You are attempting to perform an action other than Deploy on Chaincode that is not ready and you are in developer mode. Did you forget to Deploy your chaincode?
```

Unit testing this change looks overwhelming given the size of the function this is in and the lack of a mocking framework (not to mention the lack of any tests for this file at all). I am willing to try and get something in place if it is essential, otherwise I suggest testing this file should be done in another pull request.
## Checklist:

<!-- To check a box, and an 'x': [x] -->

<!-- To uncheck box, add a space: [ ] -->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff).
- [x] I have either added documentation to cover my changes or this change requires no new documentation.
- [x] I have either added unit tests to cover my changes or this change requires no new tests.
- [x] I have run [golint](https://github.com/golang/lint) and have fixed valid warnings in code I have added or modified. This tool generates false positives so you may choose to ignore some warnings. The goal is clean, consistent, and readable code.

<!-- The continuous integration build process will run [make checks](https://github.com/hyperledger/fabric/blob/master/Makefile#L22) to confirm that tests pass and that code quality meets minimum standards. You may optionally run this locally as PRs will not be accepted until they pass. -->

Signed-off-by: Julian Carrivick cjulian@au1.ibm.com

juliancarrivick-ibm committed:
Fix Issue #809
Add Error message when a user attempts an action other than deploy when they are running their own chaincode and in Developer Mode

gabre commented:
Is it possible to write a bdd test for this? It would be cool. E.g. searching the peer output for an error message like this when one has no chain code deployed.

juliancarrivick-ibm commented:
@gabre I'll give it a shot :)

juliancarrivick-ibm committed:
Add initial Logging BDD Tests, still a WIP - logic needs to be implemented

juliancarrivick-ibm commented:
I'm not going to finish this today (I've had to learn all about the wonderful world of Behaviour Driven Tests). I'm also away until next Tuesday. I'll push what I've done so far (written the tests, need to implement the logic) and hopefully finish it on Tuesday.

juliancarrivick-ibm committed:
Still a WIP - Got first 2 tests working, still need to tweak logic for final 2 tests

Implemented deploy chaincode by name in BDD step

Move other deploy steps to use refactored deploy code

Refactored test util to be able to return full container name from peer name

Log checking implemented, need to tweak logic to make tests pass

Ensure first 2 logging tests pass

juliancarrivick-ibm committed:
Get final 2 logging bdd tests working

juliancarrivick-ibm committed:
Fix CI errors when running chaincode in container

juliancarrivick-ibm committed:
Update host.sh to correspond with #2267 to allow tests to run

juliancarrivick-ibm commented:
Ready to be looked at, please note also incorporates changes from #2267 so that new tests will run.

srderson commented:
LGTM, thanks!
